OSDOCS-3048 Tech preview of KEDA in 4.11 (#42918)
openshift/openshift-docs: modules/nodes-pods-autoscaling-custom-trigger.adoc

// Module included in the following assemblies:
//
// * nodes/nodes-pods-autoscaling-custom.adoc

:_content-type: CONCEPT
[id="nodes-pods-autoscaling-custom-trigger_{context}"]
= Understanding the custom metrics autoscaler triggers

A trigger, or scaler, provide the metrics that the Horizontal Pod Autoscaler uses to scale your pods. 

The following are examples of some triggers you can use with the custom metrics autoscaler (CMA). For a full list of available triggers, see the link:https://keda.sh/docs/2.7/scalers[Keda Scalers] documentation. Because the custom metrics autoscaler is Technology Preview, Red Hat does not recommend using the triggers/scalers in production. 

You use a `ScaledObject` or `ScaledJob` to configure triggers for specific objects, as described in the sections that follow. 

[discrete]
[id="nodes-pods-autoscaling-custom-trigger-prom_{context}"]

> == Understanding the Prometheus trigger

Reviewer: We should add a paragraph or subsection here, in case users would like to scale apps based on Prometheus, that is part of OpenShift Monitoring -> specific setup is required here. I described it here: https://github.com/zroubalik/keda-openshift-examples/tree/main/prometheus/ocp-monitoring

Author: @zroubalik 
Can people use the Monitoring docs here?
https://docs.openshift.com/container-platform/4.10/monitoring/configuring-the-monitoring-stack.html#enabling-query-logging-for-thanos-querier_configuring-the-monitoring-stack

Reviewer: Yeah, this is a great reference for enabling of monitoring for user workloads (this is an prerequisite).

But then they need to:
[1. create a ServiceAccount to get a token](https://github.com/zroubalik/keda-openshift-examples/tree/main/prometheus/ocp-monitoring#2-create-a-service-account-or-reuse-existing-one-locate-assigned-token) 
[2. create a Role](https://github.com/zroubalik/keda-openshift-examples/tree/main/prometheus/ocp-monitoring#4-create-a-role-for-reading-metric-from-thanos) 
[3. add that role to the previously created Service Account](https://github.com/zroubalik/keda-openshift-examples/tree/main/prometheus/ocp-monitoring#5-add-the-role-for-reading-metrics-from-thanos-to-the-service-account).
[4. and finally reference the token (secret) in the TriggerAuthentication object used by Prometheus Scaler](https://github.com/zroubalik/keda-openshift-examples/tree/main/prometheus/ocp-monitoring#3-define-triggerauthentication-with-the-service-accounts-token)

We should cover this in the docs, becuase it is not very straightforward and we should enable users to use OpenShift Monitoring Prometheus (Thanos) instance.

Reviewer: A section like `How to scale based on metrics from OpenShift Monitoring`

Author: See: [Configuring the custom metrics autoscaler to use Prometheus Cluster Monitoring](http://file.rdu.redhat.com/~mburke/OSDOCS-3048-custom-autoscaling-keda/nodes/pods/nodes-pods-autoscaling-custom.html#nodes-pods-autoscaling-custom-prometheus-config_nodes-pods-autoscaling-custom)

Reviewer: looking good, could you please add a link to that section in here? So users reading about Prometheus trigger could quickly check this guide?

<2> Specifies the name of the Kafka topic on which Kafka is processing the offset lag.
<3> Specifies a comma-separated list of Kafka brokers to connect to.
<4> Specifies the name of the Kafka consumer group used for checking the offset on the topic and processing the related lag.
<5> Optional: Specifies the average target value to trigger scaling actions. The default is `5`.
<6> Optional: Specifies the Kafka offset reset policy for the Kafka consumer. The available values are: `latest` and `earliest`. The default is `latest.
<7> Optional: If `true`, specifies that the number of Kafka replicas can exceed the number of partitions on a topic. This allows for idle consumers. The default is `false`.
<8> Specifies how the trigger/scaler behaves when a Kafka partition does not have a valid offset. The default is `false`.
     * If `false`, the scaler keeps a single consumer for that partition.
     * If `true`, the consumers is scaled to zero for that partition.
<9> Optional: Specifies the version of your Kafka brokers. The default is `1.0.0`.

[discrete]
[id="nodes-pods-autoscaling-custom-trigger-cpu_{context}"]
== Understanding the CPU trigger

> You can scale based on CPU metrics. 

Author: How about:

> This trigger uses cluster metrics as the source for metrics.

----
<1> Specifies Kafka as the scaler/trigger type.
<2> Specifies the name of the Kafka topic on which Kafka is processing the offset lag.
<3> Specifies a comma-separated list of Kafka brokers to connect to.
<4> Specifies the name of the Kafka consumer group used for checking the offset on the topic and processing the related lag.
<5> Optional: Specifies the average target value to trigger scaling actions. The default is `5`.
<6> Optional: Specifies the Kafka offset reset policy for the Kafka consumer. The available values are: `latest` and `earliest`. The default is `latest.
<7> Optional: If `true`, specifies that the number of Kafka replicas can exceed the number of partitions on a topic. This allows for idle consumers. The default is `false`.
<8> Specifies how the trigger/scaler behaves when a Kafka partition does not have a valid offset. The default is `false`.
     * If `false`, the scaler keeps a single consumer for that partition.
     * If `true`, the consumers is scaled to zero for that partition.
<9> Optional: Specifies the version of your Kafka brokers. The default is `1.0.0`.

[discrete]
[id="nodes-pods-autoscaling-custom-trigger-cpu_{context}"]

> == Understanding the CPU trigger

Reviewer: We should note that this trigger can't be used in ScaledJobs.

You can scale based on CPU metrics. This trigger uses cluster metrics as the source for metrics.

.Example scaled object with a CPU scaler
[source,yaml,options="nowrap"]
----
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: cpu-scaledobject
  namespace: my-namespace
spec:

 ...

  triggers:
  - type: cpu <1>
    metricType: Utilization <2>
    metadata:
      value: "60" <3>
----
<1> Specifies CPU as the scaler/trigger type.
<2> Specifies the type of metric to use, either `Utilization` or `AverageValue`.
<3> Specifies the value to trigger scaling actions for:
* When using `Utilization`, the target value is the average of the resource metrics across all relevant pods, represented as a percentage of the requested value of the resource for the pods.
* When using `AverageValue`, the target value is the average of the metrics across all relevant pods.

[discrete]
[id="nodes-pods-autoscaling-custom-trigger-memory_{context}"]

> == Understanding the memory trigger

Reviewer: We should note that this trigger can't be used in ScaledJobs.

// Module included in the following assemblies:
//
// * nodes/nodes-pods-autoscaling-custom.adoc

:_content-type: CONCEPT
[id="nodes-pods-autoscaling-custom-trigger_{context}"]
= Understanding the custom metrics autoscaler triggers

Triggers, also known as scalers, provide the metrics that the Custom Metrics Autoscaler Operator uses to scale your pods. You can specify a single trigger or multiple triggers. When using multiple triggers, the scaling is based on the greatest value from all the triggers.

> This section contains examples of some triggers you can use with the custom metrics autoscaler (CMA). For a full list of available triggers, see the link:https://keda.sh/docs/2.7/scalers[KEDA Scalers] documentation. Because the custom metrics autoscaler is Technology Preview, Red Hat does not recommend using the triggers/scalers in production. 

Reviewer 2: Is Red Hat supporting all of the Scalers listed by the upstream? IE: if a customer finds a bug with one of these scalers (or the upstream adds a scaler in the middle of a 4.Y version's lifecycle) can a customer take advantage of it? Can they get patches and security updates for the scalers?

If the answer is no; to this then we shouldn't like to the upstream here; but instead we should list what Red Hat is supporting.

All in all I don't have an issue linking to the upstream here; but we have to clarify this paragraph with a support clause / call out that said (more or less): Red Hat provides and only supports the ability to use scalers; the source / support for each individual scaler is offered by the provider of at scaler (please see the provider for support options), etc, etc.

// Module included in the following assemblies:
//
// * nodes/nodes-pods-autoscaling-custom.adoc

:_content-type: CONCEPT
[id="nodes-pods-autoscaling-custom-trigger_{context}"]
= Understanding the custom metrics autoscaler triggers

Triggers, also known as scalers, provide the metrics that the Custom Metrics Autoscaler Operator uses to scale your pods. You can specify a single or multiple triggers. When using multiple triggers, the scaling is based on the greatest value from all the triggers.

> The following are examples of some triggers you can use with the custom metrics autoscaler (CMA). For a full list of available triggers, see the link:https://keda.sh/docs/2.7/scalers[KEDA Scalers] documentation. Because the custom metrics autoscaler is Technology Preview, Red Hat does not recommend using the triggers/scalers in production. 

Author: They are further down. I will change this wording.

You use a `ScaledObject` or `ScaledJob` custom resource to configure triggers for specific objects, as described in the sections that follow. 

[id="nodes-pods-autoscaling-custom-trigger-prom_{context}"]
== Understanding the Prometheus trigger

Scale applications based on Prometheus metrics. See *Additional references* for information on the configurations required to use the {product-title} Prometheus Cluster Monitoring as a source for metrics. 

[NOTE]
====

> If Prometheus is taking metrics from the application that the custom metric autoscaler is scaling, do not set the minimum replicas to `0` in the custom resource. If there were no application pods, the custom metric autoscaler would not have any metrics to scale on.

Reviewer 3: ```suggestion
If Prometheus is taking metrics from the application that the custom metric autoscaler is scaling, do not set the minimum replicas to `0` in the custom resource. If there are no application pods, the custom metric autoscaler does not have any metrics to scale on.
```
Does this mean the same thing?

====

.Example scaled object with a Prometheus target
[source,yaml,options="nowrap"]
----
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: prom-scaledobject
  namespace: my-namespace
spec:

> 

Reviewer 3: I'd remove the extra blank lines

// Module included in the following assemblies:
//
// * nodes/nodes-pods-autoscaling-custom.adoc

:_content-type: CONCEPT
[id="nodes-pods-autoscaling-custom-trigger_{context}"]
= Understanding the custom metrics autoscaler triggers

Triggers, also known as scalers, provide the metrics that the Custom Metrics Autoscaler Operator uses to scale your pods. 

[NOTE]
====

> {product-title} curently supports only the Prometheus trigger, which can use the installed {product-title} monitoring or an external Prometheus server as the metrics source.  

Reviewer: `{product-title} curently supports only the Prometheus trigger` -> this is rendered as  `OpenShift Container Platform curently supports only the Prometheus trigger`

But should probably be:
`The the custom metrics autoscaler curently supports only the Prometheus trigger`

Scale applications based on Prometheus metrics. See _Additional resources_ for information on the configurations required to use the {product-title} monitoring as a source for metrics. 

[NOTE]
====
If Prometheus is taking metrics from the application that the custom metrics autoscaler is scaling, do not set the minimum replicas to `0` in the custom resource. If there are no application pods, the custom metrics autoscaler does not have any metrics to scale on.
====

.Example scaled object with a Prometheus target
[source,yaml,options="nowrap"]
----
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: prom-scaledobject
  namespace: my-namespace
spec:
 ...
  triggers:
  - type: prometheus <1>
    metadata:
      serverAddress: https://thanos-querier.openshift-monitoring.svc.cluster.local:9092 <2>
      namespace: kedatest <3>
      metricName: http_requests_total <4>
      threshold: '5' <5>
      query: sum(rate(http_requests_total{job="test-app"}[1m])) <6>
      authModes: "basic" <7>
----
<1> Specifies Prometheus as the scaler/trigger type.
<2> Specifies the address of the Prometheus server. This example uses  {product-title} monitoring.

> <3> Optional: Specifies the namespace of the object you want to scale.

Reviewer: ```suggestion
<3> Optional: Specifies the namespace of the object you want to scale. This property must be set if {product-title} monitoring Prometheus instance is being used.
```
